xbmc/cores/VideoPlayer/VideoRenderers/VideoShaders/ConversionMatrix.h

```diff
@@ -18,34 +20,85 @@ template <unsigned Order>
 class CMatrix
 {
 public:
-  CMatrix(float (&src)[Order][Order]);
-  CMatrix(float (&src)[Order-1][Order-1]);
-  CMatrix& operator=(const CMatrix& src);
-  CMatrix& operator=(const float (&src)[Order-1][Order-1]);
+  CMatrix() = default;
+  CMatrix(CMatrix<Order - 1>& other);
+  CMatrix(std::array<std::array<float, Order>, Order>& other) { m_mat = other; }
+  CMatrix(std::array<std::array<float, Order - 1>, Order - 1>& other);
   virtual ~CMatrix() = default;
 
-  float (&Get())[Order][Order];
-  CMatrix Invert();
+  virtual CMatrix operator*(const std::array<std::array<float, Order>, Order>& other);
+
   CMatrix operator*(const CMatrix& other);
   CMatrix operator*=(const CMatrix& other);
-  virtual CMatrix operator*(const float (&other)[Order][Order]);
+
+  CMatrix& operator=(const std::array<std::array<float, Order - 1>, Order - 1>& other);
+
+  bool operator==(const CMatrix<Order>& other) const
+  {
+    for (unsigned i = 0; i < Order; ++i)
+    {
+      for (unsigned j = 0; j < Order; ++j)
+      {
+        if (m_mat[i][j] == other.m_mat[i][j])
+          continue;
+
+        // some floating point comparisions should be done by checking if the difference is within a tolerance
+        if (std::abs(m_mat[i][j] - other.m_mat[i][j]) <=
+            (std::max(std::abs(other.m_mat[i][j]), std::abs(m_mat[i][j])) * 1e-2))
+          continue;
+
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  std::array<float, Order>& operator[](int index) { return m_mat[index]; }
+
+  std::array<std::array<float, Order>, Order>& Get();
+
+  CMatrix Invert();
+
+  float* ToRaw()
```

Comment:
What is this function doing?

Reply:
It's getting the raw array of arrays, I'm not sure if there is a better way to handle this or not. The problem is that there is some padding in `std::array` (as far as I'm aware) so doing something like m_mat.data() isn't exactly possible as you need the subarray also. Maybe you have a suggestion about how to improve this?

Reply:
Hrm I still don't understand what the need of the loop is. Since there's no `break` and you reset `raw` on every iteration, it should be completely equivalent to just `raw = &m_mat[Order - 1][Order - 1] + 1 - (Order * Order)`. Which should be equivalent to `&m_mat[0][0]`, shouldn't it? Also, there is no padding in front of the elements as far as I am aware. Quote: "This container is an aggregate type with the same semantics as a struct holding a C-style array T[N] as its only non-static data member." (https://en.cppreference.com/w/cpp/container/array). I'm not sure whether there may be padding between the elements, but this function wouldn't save you from that either. So I think you can actually just return `m_mat.data()`. Or `&m_mat[0][0]` if you want.

Reply:
I'll give it a try

Reply:
`&m_mat[0][0]` seems to work, m_mat.data() doesn't compile

Reply:
makes sense, guess it would have to be `m_mat.data()->data()`, which looks quite ugly. better say with `&m_mat[0][0]`